### PR TITLE
feat(sync): publish DeepSeek V4 Flash support and training fixes

### PR DESCRIPTION
**Pull request by SigureMo**
## What

Publish eight changes from the development branch: DeepSeek V4 training and SFT support, shared-memory model loading fixes, rollout-resume synchronization, INT4 QAT and expert-gradient corrections, DeepEyes cleanup, startup failure propagation, and training-script guidance.

The follow-up fix also preserves the NPU Megatron constructor contract and keeps all six indexer replay CPU regressions active without importing the distributed runtime during test collection.

## Why

The public branch is missing these development changes. Startup failures should exit promptly instead of leaving services running, and restoring model weights must preserve tokenizer metadata that active readers still use.

## How

- Add DeepSeek V4 model conversion, CP/indexer replay, SFT chat-template handling, and example launch scripts.
- Make shared-memory model cache restoration transactional; synchronize training backends before resuming rollout.
- Remove duplicate INT4 fake-QAT injection, disable incompatible gradient accumulation fusion in the INT4 recipes, and include expert gradients in the TP norm calculation.
- Remove the obsolete DeepEyes R3 processor patch and propagate service initialization failures through bounded cleanup.
- Construct packed-sequence parameters using the shared constructor fields, then attach padding metadata. This addresses the [NPU compatibility review](https://github.com/redai-studio/Relax/pull/343#discussion_r4056341957) while retaining newer Megatron padding semantics. A TODO marks this workaround for removal once the NPU Megatron pin supports the constructor argument.
- Replay the eight source commits individually with provenance trailers. Preserve the public algorithm registry, optional-dependency test isolation, KLX fixes, and documentation anchors.

Please use **Rebase and merge** to retain the linear history and source commit trailers. The internal mirror will be updated from the resulting public `main` after this PR is merged.

## Testing

All seven GitHub checks passed for `141e9ffef6c1bdccf007dc3ce71f360c4059c268`.

- [x] Full pre-commit checks, including gitleaks, passed before amending the follow-up commit.
- [x] **34 existing focused CPU regressions passed**, covering SFT data, indexer replay, streaming iterators, and train data fields.
- [x] [Python CI](https://github.com/redai-studio/Relax/actions/runs/35501048464): Python 3.10, 3.11, and 3.12 passed.
- [x] [GPU unit tests on 4 H20 GPUs](https://github.com/redai-studio/Relax/actions/runs/35501048465) passed.
- [x] [GPU integration tests](https://github.com/redai-studio/Relax/actions/runs/35501048481): Qwen3-4B async on 4 GPUs and Qwen3-VL-4B on 2 GPUs passed.
- [ ] NPU training, multi-node training, and container builds were not run as part of this compatibility validation.

The constructor compatibility fix requires no dependency or Dockerfile version change. Full NPU training validation remains outstanding.

## Type of Change

- [x] Bug fix
- [x] New feature
- [x] Documentation update
- [x] Refactoring
- [x] Build changes

### CODE REVIEW

File: relax/backends/megatron/data.py [hunk 1]
```diff
@@ -473,13 +588,17 @@ def get_batch(
...
+            max_seqlen = max(end - start for start, end in zip(cu_seqlens_cpu, cu_seqlens_cpu[1:]))
         packed_seq_params = PackedSeqParams(
             cu_seqlens_q=cu_seqlens,
             cu_seqlens_kv=cu_seqlens,
+            cu_seqlens_q_padded=cu_seqlens_padded if allgather_cp else None,
+            cu_seqlens_kv_padded=cu_seqlens_padded if allgather_cp else None,
             max_seqlen_q=max_seqlen,
             max_seqlen_kv=max_seqlen,
             qkv_format="thd",
+            pad_between_seqs=pad != 0 if allgather_cp else None,
```
[the remark this answers is not in the document]
Comment (SigureMo) [reply relax/backends/megatron/data.py]: @rai-studio-bot 是指从 `docker/Dockerfile.npu` 构建的 NPU 镜像吗？是说这里需要更新 dockerfile 还是单纯更新镜像？
